File: core/emergency_handler.py
from __future__ import annotations

from typing import Tuple

from core.voice_engine import VoiceEngine

try:
    from jnius import autoclass
except Exception:  # pragma: no cover
    autoclass = None

Coordinate = Tuple[float, float]


class EmergencyHandler:
    """Handles long-press emergency flow and SMS dispatch with GPS coordinates."""

    def __init__(self, voice_engine: VoiceEngine):
        self.voice_engine = voice_engine

    def activate(self, location: Coordinate, contact_number: str) -> bool:
        lat, lon = location
        message = f"Emergency alert from BlindNav AI. Location: https://maps.google.com/?q={lat:.6f},{lon:.6f}"
        sent = self._send_sms(contact_number, message)

        if sent:
            self.voice_engine.speak_alert("Emergency message sent with your GPS coordinates.", emergency=True)
        else:
            self.voice_engine.speak_alert("Emergency mode activated, but SMS could not be sent.", emergency=True)
        return sent

    @staticmethod
    def _send_sms(phone: str, message: str) -> bool:
        if autoclass is None:
            return False
        try:
            SmsManager = autoclass("android.telephony.SmsManager")
            sms_manager = SmsManager.getDefault()
            sms_manager.sendTextMessage(phone, None, message, None, None)
            return True
        except Exception:
            return False

File: app/src/main/java/com/blindnavai/emergency/EmergencyHandler.kt
package com.blindnavai.emergency

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.telephony.SmsManager
import androidx.core.content.ContextCompat
import com.blindnavai.voice.VoiceManager
import com.google.android.gms.location.FusedLocationProviderClient

class EmergencyHandler(
    private val context: Context,
    private val fusedLocationProviderClient: FusedLocationProviderClient,
    private val voiceManager: VoiceManager,
    private val emergencyNumber: String = "112"
) {
    fun trigger(lastKnownLocation: Location?) {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.SEND_SMS) != PackageManager.PERMISSION_GRANTED) {
            voiceManager.announceHazard("Emergency SMS permission is not granted")
            return
        }

        if (lastKnownLocation != null) {
            sendSms(lastKnownLocation.latitude, lastKnownLocation.longitude)
        } else {
            fusedLocationProviderClient.lastLocation.addOnSuccessListener { location ->
                if (location != null) {
                    sendSms(location.latitude, location.longitude)
                } else {
                    voiceManager.announceHazard("Unable to get location for emergency message")
                }
            }.addOnFailureListener {
                voiceManager.announceHazard("Emergency mode failed to retrieve location")
            }
        }
    }

    private fun sendSms(lat: Double, lon: Double) {
        val mapsLink = "https://maps.google.com/?q=$lat,$lon"
        val message = "BlindNav AI emergency alert. Current location: $lat,$lon. Map: $mapsLink"
        SmsManager.getDefault().sendTextMessage(emergencyNumber, null, message, null, null)
        voiceManager.announceHazard("Emergency message sent with your current coordinates")
    }
}
